MapboxNavigation/RouteMapViewController.swift

```diff
@@ -254,55 +259,133 @@ extension RouteMapViewController: NavigationMapViewDelegate {
                 newCoordinate = coordinate
             }
         }
         
+        // Add current way name to UI
+        if let style = mapView.style, recenterButton.isHidden {
+            let streetsLanguages = ["zh", "ru", "fr", "es", "en", "de"]
+            let roadLabelLayerIdentifier = "roadLabelLayer"
+            let streetsSources = style.sources.flatMap {
+                $0 as? MGLVectorSource
+                }.filter {
+                    $0.isMapboxStreets
+            }
+            assert(!streetsSources.isEmpty, "Style must contain the source `mapbox.mapbox-streets-v7`")
+            
+            if let mapboxSteetsSource = streetsSources.first, style.layer(withIdentifier: roadLabelLayerIdentifier) == nil {
+                let streetLabelLayer = MGLLineStyleLayer(identifier: roadLabelLayerIdentifier, source: mapboxSteetsSource)
+                streetLabelLayer.sourceLayerIdentifier = "road_label"
+                
+                // If the opacity is set to 0, the feature will be ignored in `mapView.visibleFeatures()`
+                streetLabelLayer.lineOpacity = MGLStyleValue(rawValue: 1)
+                streetLabelLayer.lineWidth = MGLStyleValue(rawValue: 20)
+                streetLabelLayer.lineColor = MGLStyleValue(rawValue: .white)
+                style.insertLayer(streetLabelLayer, at: 0)
+            }
+            
+            let userPuck = mapView.convert(newCoordinate, toPointTo: mapView)
+            let features = mapView.visibleFeatures(at: userPuck, styleLayerIdentifiers: Set([roadLabelLayerIdentifier]))
+            var smallestLabelDistance = Double.infinity
+            var currentName: String?
+            
+            for feature in features {
+                var allLines: [MGLPolyline] = []
+                
+                if let line = feature as? MGLPolylineFeature {
+                    allLines.append(line)
+                } else if let lines = feature as? MGLMultiPolylineFeature {
+                    allLines = lines.polylines
+                }
+                
+                for line in allLines {
+                    let featureCoordinates =  Array(UnsafeBufferPointer(start: line.coordinates, count: Int(line.pointCount)))
+                    let slicedLine = polyline(along: coordinates, from: newCoordinate)
+                    let featureSlice = polyline(along: featureCoordinates, from: newCoordinate)
+                    // todo: account for -10
+                    
+                    let lookAheadDistance:CLLocationDistance = 10
+                    
+                    let pointAheadFeature = coordinate(at: lookAheadDistance, fromStartOf: featureSlice)!
+                    let pointAheadUser = coordinate(at: lookAheadDistance, fromStartOf: slicedLine)!
+                    let pointBehindFeature = coordinate(at: -lookAheadDistance, fromStartOf: featureSlice)!
+                    let pointBehindUser = coordinate(at: -lookAheadDistance, fromStartOf: slicedLine)!
```

Comment:
Correct, `coordinate(at:fromStartOf:)` doesn’t support negative distances. We could either create a `coordinate(at:fromEndOf:)` method or pass in the `reversed()` coordinate array.